src/v/storage/tiered_storage.cc

```diff
@@ -0,0 +1,516 @@
+        } else if (prop.retention_time.has_value()) {
+            return time_bound_deletion_parameters{prop.retention_time.value()};
+        }
+    }
+    return std::monostate();
+}
+
+ss::future<partition_downloader::offset_map_t>
+partition_downloader::build_offset_map(const recovery_material& mat) {
+    // We have multiple versions of the same partition here, some segments
+    // may overlap so we need to deduplicate. Also, to take retention into
+    // account.
+    offset_map_t offset_map;
+    for (const auto& p : mat.paths) {
+        auto manifest = co_await download_manifest(p);
+        for (const auto& segm : manifest) {
+            if (offset_map.contains(segm.second.base_offset)) {
+                auto committed = offset_map.at(segm.second.base_offset)
+                                   .meta.committed_offset;
+                if (committed > segm.second.committed_offset) {
+                    continue;
+                }
+            }
+            auto path = manifest.get_remote_segment_path(segm.first);
+            offset_map.insert_or_assign(
+              segm.second.base_offset,
+              segment{.full_path = path().native(), .meta = segm.second});
+        }
+    }
+    co_return std::move(offset_map);
+}
+
+// entry point for the whole thing
+ss::future<> partition_downloader::download_log(
+  const cloud_storage::remote_manifest_path& manifest_key) {
+    auto prefix = std::filesystem::path(_ntpc.work_directory());
+    auto retention = get_retention_policy(_ntpc.get_overrides());
+    vlog(
+      _ctxlog.info,
+      "The target path: {}, ntp-config revision: {}, retention: {}",
+      prefix,
+      _ntpc.get_revision(),
+      retention);
+    auto mat = co_await find_recovery_material(manifest_key);
+    auto offset_map = co_await build_offset_map(mat);
+    cloud_storage::manifest target(_ntpc.ntp(), _ntpc.get_revision());
+    for (const auto& kv : offset_map) {
+        // Original manifests contain short names (e.g. 1029-4-v1.log).
+        // This is because they belong to the same revision and the details are
+        // encoded in the manifest itself.
+        // To create a compound manifest we need to add full names (e.g.
+        // 6fab5988/kafka/redpanda-test/5_6/80651-2-v1.log). Otherwise the
+        // information in the manifest won't be suffecient.
+        target.add(
+          cloud_storage::segment_name(kv.second.full_path), kv.second.meta);
+    }
+    if (recovery_logger.is_enabled(ss::log_level::debug)) {
+        std::stringstream ostr;
+        target.serialize(ostr);
+        vlog(_ctxlog.debug, "Generated partition manifest: {}", ostr.str());
+    }
+    // Here the partition manifest 'target' may contain segments
+    // that have different revision ids inside the path.
+    if (target.size() == 0) {
+        vlog(
+          _ctxlog.error,
+          "No segments found. Empty partition manifest generated.");
+        throw missing_partition_exception(_ntpc);
+    }
```

Comment:
Before attempting size-bound or time-bound restoration of the topic, there should be a pre-check for available disk space in the redpanda data directory.  During disaster recovery, a user may well be restoring to a physically different machine than the original writer of the topic, with different drives.

Then if enough space isn't available, we could log an error suggesting that they use a size-bound setting for the topic to enable restore.